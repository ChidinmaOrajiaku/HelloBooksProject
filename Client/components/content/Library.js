import React from 'react';
import map from 'lodash/map';
import axios from 'axios';
import PropTypes from 'prop-types';
import { connect } from 'react-redux';
import { store } from '../../index';
import { borrowRequest, getRequest } from '../../actions/booksAction';

class Library extends React.Component {
    constructor (props) {
        super(props);
        this.state = {
          user: (store.getState()).auth.user.id,
          title: '',
          author: '',
          category: '',
          image: '',
          review: '',
          booksId: '',
          data: []
        }
        this.onBorrowBooks = this.onBorrowBooks.bind(this);
        this.handleChange = this.handleChange.bind(this);
      }

      handleChange(event) {
        this.setState({[event.target.id]: event.target.value})
      }
       
      onBorrowBooks(event) {
        this.props.borrowRequest(this.state.user,this.state).then(
          () => {Materialize.toast(this.state.profileResponse, 2000, 'teal rounded')},
          (errors) =>{
            Materialize.toast(errors.response.data.message, 2000, 'red accent-3 rounded')
         this.setState({ errors: errors.response.data.message  })
          }
        )
      }
      
      componentWillMount() {
        axios.get('/api/v1/users/books').then((res) => {
          localStorage.getItem('jwtToken');
          this.setState({ data: res.data})
        });
      }
       
    render() {

    $(document).ready(function(){
      $('.tooltipped').tooltip({delay: 50});
      $('.scrollspy').scrollSpy()
    });
    
    const { data } = this.state
    const { borrowRequest } = this.props
  return (
      <div className= "library">
        <div className=""> <NavigationBar /> </div>
       <div className="row">
        <div className="col s12 m9 l10">
            {/* fictional */}
          <div id="fiction" className="section scrollspy">
          <h1 className = "libraryHeading"> Fiction </h1>
             <div className="row">
             {this.state.data.map ((books, index) =>
               <div className="col s3 " key={index}>
                  <div className="card hoverable">
                     <div className="card-image">
                         <img id="image" value={ books.image } onChange={ this.handleChange } src= { books.image }/>
                         <button value= { books.id } onChange={this.handleChange} className="btn tooltipped btn-floating halfway-fab waves-effect waves-light teal" data-position="bottom" data-delay="50" data-tooltip="Hi! Click to borrow" onClick= { this.onBorrowBooks } type="submit" name="action" ><i className="material-icons">add</i></button>
                     </div>
                      <div className="card-content">
                         <p>{books.review}</p>
                      </div>
                  </div>
             </div>   
          )
          }
          </div>
          </div>
    </div>
    <div className="col hide-on-small-only m3 l2">
      <ul className="section table-of-contents">
        <li><a href="#fiction">Fiction</a></li>
        <li><a href="#educational">Educational</a></li>
        <li><a href="#technology">Technology</a></li>
        <li><a href="#motivational">Motivational</a></li>
        <li><a href="#general">General</a></li>
      </ul>
    </div>
  </div>
  <div className='page'>
  <ul className="pagination">
  <li className="disabled"><a href="#!"><i className="material-icons">chevron_left</i></a></li>
  <li className="active"><a href="#!">1</a></li>
  <li className="waves-effect"><a href="#!">2</a></li>
  <li className="waves-effect"><a href="#!">3</a></li>
  <li className="waves-effect"><a href="#!">4</a></li>
  <li className="waves-effect"><a href="#!">5</a></li>
  <li className="waves-effect"><a href="#!"><i className="material-icons">chevron_right</i></a></li>
</ul>
</div>
</div>
   );
};
}


Library.propTypes = {
  borrowRequest: PropTypes.func.isRequired,
  getRequest: PropTypes.func.isRequired
}


export default connect(null, {borrowRequest, getRequest}) (Library);
