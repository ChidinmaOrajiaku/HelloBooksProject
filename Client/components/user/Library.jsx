import React from 'react';
import ReactDOM from 'react-dom';
import { connect } from 'react-redux';
import { Link } from 'react-router-dom';
import Footer from '../Footer';
import NavigationBar from '../NavigationBar';
import { getRequest } from '../../actions/getAllBooks';
import { borrowRequest } from '../../actions/borrowBooks';


/**
 * @class Books
 */
class Library extends React.Component {
  /**
     * @constructor
     * @param {object} props 
     */
  constructor(props) {
    super(props);
    this.state = {
      getAllBooks: [],
      loading: true,
      booksId: ''
    };
    this.handleBorrow = this.handleBorrow.bind(this);
  }

  /**
   * 
   * @returns {object} response object
   * @memberof Library
   */
  componentDidMount() {
    this.props.getRequest();
    $('select').material_select();
    $('select').change(event => this.handleChange(event));
    $('.tooltipped').tooltip();
  }

  /**
   * Receives next props an sets the state
   * @returns {object} response object
   * @param {nextProps} nextProps 
   * @memberof Library
   */
  componentWillReceiveProps(nextProps) {
    this.setState({
      loading: false,
      getAllBooks: nextProps.getAllBooksData,
    });
  }

  /**
   * Handles borrow books action
   * @returns {object} response object
   * @param {event} event 
   * @memberof Books
   */
  handleBorrow(event) {
    event.preventDefault();
    localStorage.setItem('bookId', event.target.value);
    this.setState({
      booksId: localStorage.getItem('bookId')
    }),
    setTimeout(() => {
      this.props.borrowRequest(this.props.usersId, this.state).then(() => {
        Materialize.toast('Succesfully borrowed', 2000, 'teal rounded');
      }).catch(() => {
        Materialize.toast('Sorry! You cannot borrow this book',
          2000, 'red rounded');
      });
    });
  }

  /**
     * 
     * React element mark up
     * @returns {object} response object
     * @memberof Library
     */
  render() {
    const books = this.state.loading ? <div><p>Loading...</p></div> :
      <div className=" lib row">
        { Object.keys(this.state.getAllBooks).map(key =>
          <div className="col s12 m4 push-m2" key={key}>
            <div className="card">
              <div className="card-image waves-effect waves-block waves-light">
                <img className="activator"
                  src={this.state.getAllBooks[key].image }
                />
              </div>
              <div className="card-content">
                <span className="card-title activator teal-text text-darken-3">
                  { this.state.getAllBooks[key].title }
                  <i className="material-icons right tooltipped"
                    data-position="bottom" data-delay="50" data-tooltip="More.."
                  >
                more_vert</i>
                </span>
              </div>
              <div className="card-reveal">
                <span className="card-title grey-text text-darken-4">
                  Book Details<i className="material-icons right">close</i>
                </span>
                <p> <b className="teal-text text-darken-3">Title:
                </b> { this.state.getAllBooks[key].title }</p>
                <p> <b className="teal-text text-darken-3">Author</b>:
                  { this.state.getAllBooks[key].author }</p>
                <p> <b className="teal-text text-darken-3">Category</b>:
                  { this.state.getAllBooks[key].category }</p>
                <p> <b className="teal-text text-darken-3">Review</b>:
                  { this.state.getAllBooks[key].review }</p>
                <button className="waves-effect waves-light btn borrowButton"
                  value={ this.state.getAllBooks[key].id }
                  onClick={this.handleBorrow}>Borrow </button>
              </div>
            </div>
          </div>
        )}
      </div>;

    return (
      <div className="library row container">
        <div> <NavigationBar /> </div>
        <h4 className="col m10 offset-m2"> USER LIBRARY </h4>
        {books}
        <div> <Footer /></div>
      </div>
    );
  }
}

const mapStateToProps = state => (
  {
    usersId: state.auth.user.id,
    getAllBooksData: state.getAllBooks[0].response,
  }
);

export default connect(mapStateToProps,
  { getRequest, borrowRequest })(Library);
