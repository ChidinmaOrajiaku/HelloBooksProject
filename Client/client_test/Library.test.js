import React from 'react';
import $ from 'jquery';
import toJson from 'enzyme-to-json';
import { configure, shallow, mount } from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';
import { Library, mapStateToProps } from '../components/user/Library.jsx';

configure({ adapter: new Adapter() });

describe('<Library />', () => {
  const props = {
    getRequest: jest.fn()
  };
  let mountedComponent;
  const libraryItem = () => {
    if (!mountedComponent) {
      mountedComponent = shallow(<Library {...props}/>);
    }
    return mountedComponent;
  };

  it('should call handleReturn method', () => {
    const spy = jest.spyOn(libraryItem().instance(), 'handleBorrow');
    const event = {
      preventDefault: jest.fn(),
      target: {
        value: 'abd',
        dataset: {
          index: 1
        }
      }
    };
    libraryItem().instance().handleBorrow(event);
    expect(spy).toHaveBeenCalled();
  });

  it('should call componentWillReceiveProps method', () => {
    const nextProps = {
      getAllBooksData: [],
    };
    const spy = jest.spyOn(Library.prototype, 'componentWillReceiveProps');
    shallow(<Library {...props} componentWillReceiveProps={spy}/>)
      .instance().componentWillReceiveProps(nextProps);
    expect(spy).toHaveBeenCalledTimes(1);
  });

  it('ensures that mapStateToProps dispatches the specified actions', () => {
    const state = {
      auth: {
        user: {
          id: 1
        }
      },
      getAllBooks: ['1', '2'],
      borrowBooks: ['1', '2'],
    };
    expect(mapStateToProps(state).usersId).toExist;
    expect(mapStateToProps(state).getAllBooksData).toExist;
    expect(mapStateToProps(state).borrowBooksData).toExist;
  });

  it('should match snapshot test', () => {
    const component = libraryItem();
    const tree = toJson(component);
    expect(tree).toMatchSnapshot();
  });
});
