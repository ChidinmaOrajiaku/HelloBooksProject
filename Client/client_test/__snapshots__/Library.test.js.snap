// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Library /> should match snapshot test 1`] = `
<div
  className="library row"
>
  <div
    className="container"
  >
    <div>
       
      <Connect(withRouter(NavigationBar)) />
       
    </div>
    <h4
      className="col s12 m10 offset-m2 white-text"
    >
       USER LIBRARY 
    </h4>
    <div>
      <p>
        Loading...
      </p>
    </div>
    <div>
       
      <Component />
    </div>
  </div>
</div>
`;
